### PR TITLE
AbortController ponyfill

### PR DESCRIPTION
**Pull request by stereobooster**
Related:
- https://github.com/developit/unfetch/issues/54
- https://github.com/developit/unfetch/pull/68

### CODE REVIEW

File: src/AbortController.mjs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export default function() {
+	this.signal = { onabort: () => {} };
```
Comment (joaovieira) [inline src/AbortController.mjs]: signal has an `aborted` flag indicating it's been aborted or not, which can be handy:

https://dom.spec.whatwg.org/#abortsignal-aborted-flag
https://github.com/mo/abortcontroller-polyfill/blob/master/src/abortcontroller.js#L42
https://github.com/mysticatea/abort-controller/blob/master/src/abort-signal.mjs#L28

File: src/index.mjs [hunk 2]
```diff
@@ -17,6 +17,9 @@ export default function(url, options) {
 
 		request.onerror = reject;
 
+		if (options.signal) options.signal.onabort = () => { request.abort(); }
+		request.onabort = () => reject(new DOMException('The user aborted a request.'));
```
[the remark this answers is not in the document]
Comment (joaovieira) [reply src/index.mjs]: per the spec:

```js
reject(new DOMException('Aborted', 'AbortError'));
```

https://dom.spec.whatwg.org/#aborting-ongoing-activities
https://dom.spec.whatwg.org/#aborting-ongoing-activities-example
https://github.com/github/fetch/blob/master/fetch.js#L446

File: src/index.mjs [hunk 3]
```diff
@@ -17,6 +17,9 @@ export default function(url, options) {
 
 		request.onerror = reject;
 
+		if (options.signal) options.signal.onabort = () => { request.abort(); }
```
Comment (joaovieira) [inline src/index.mjs]: if signal is already aborted it should reject promise, first thing:

https://dom.spec.whatwg.org/#aborting-ongoing-activities-spec-example
https://github.com/github/fetch/blob/master/fetch.js#L445

**Comment by stereobooster** [comment]
Thanks for links to fetch spec. 

I got the idea of passing aborted signal to the fetch, but I wonder what the use case? In what kind of API you would create N-signals and attach to fetch later? I create signal right before sending request and save it until I got response, to be able to cancel if I need to create another fetch (and only one allowed at the same time).

Should be trivial to implement, I just wonder do we need it or not

File: src/index.mjs (hunk 2 above)
[the remark this answers is not in the document]
Comment (joaovieira) [reply src/index.mjs]: Also, don't think `DOMException` is globally available:

https://developer.mozilla.org/en-US/docs/Web/API/DOMException
https://github.com/github/fetch/pull/592/files#diff-37e6c8c3d38d10cb0fc896766eb1b692R427

**Comment by joaovieira** [comment]
@stereobooster I also didn't have that use case yet. But I can think of using it as a hand-rolled circuit-breaking system where you use a signal to reject all requests going to a troubled/struggling downstream service. 

Anyway, many people will use `unfetch` with `isomorphic-unfetch` so having consistent functionality with `node-fetch` (and thus, the spec) sounds right. Specially if it's easy :)

**Pull request closed without merging**